### DISCUSSION TITLE
Fortran options when compiling UMAT code to use with tensor mechanics

### DISCUSSION THREAD

**Discussion by ngrilli** [idaholab/moose#19560] [Q&A Modules: Solid mechanics] [answered]
Dear MOOSE users,

I have recently tried the UMAT plugin.
It works well but I noticed that to compile properly my UMAT with make command in my MOOSE App folder,
I need to do some modifications compared with the UMAT I use for Abaqus.

Specifically, I notice that I need to shorten some lines by going to new line with +,
this makes me think that fortran 77 standard is followed.

Also, fortran modules are not compiled by the make command, I need to compile them separately using:
gfortran -c name_of_module.f
to generate the .mod files

The UMAT that works well is this one:
https://github.com/ngrilli/c_pfor_am/tree/main/test/plugins

but I had to make many changes compared with the original one here:
https://github.com/TarletonGroup/CrystalPlasticity

Is there a way to set gfortran options when compiling the UMAT?
For instance I would like to have the free form option and compiling with f90 standards. 

Thank you very much in advance.

Best Regards,
Nicolò

**Comment by GiudGiud** [comment]
@dschwen

**Comment by ngrilli** [comment]
@GiudGiud  @dschwen  Thank you very much in advance for the answer.

I also want to point out that in the UMAT and Abaqus mesh convention, the input variables NOEL and NPT,
which are the element number and integration point number,
start from 1 and not from 0, as in the MOOSE convention.

In the current implementation of AbaqusUMATStress this is not considered,
therefore I modified my UMAT and use NOEL+1 and NPT+1

I think this could be easily fixed in the AbaqusUMATStress
and would lead to full compatibility with UMATs, like mine, that use NOEL and NPT for some of the calculations.

Best Regards,
Nicolò

**Comment by dschwen** [comment, marked as answer]
Thanks Nicolò, the 1 based indexing is a valuable comment. I'll change that. Hopefully it won't screw things up for existing users. 

I wrote a system for compiling multi file plugins, but it does not deal with .mod files yet. An easy fix. Look at 
https://github.com/idaholab/moose/tree/next/modules/tensor_mechanics/examples/umat_build
for an example Makefile. The example is C++, but this should work for fortran (not .mod yet) files. I'll fix that today.

**Reply by dschwen** [reply]
Here is a PR addressing your points: https://github.com/idaholab/moose/pull/19593

**Comment by ngrilli** [comment]
@dschwen Thank you very much, now I understand how to compile multiple files.

What about the free form fortran option?
Is there an option I can put in the Makefile to tell the compiler that I want free form fortran?

Right now I need to truncate my entire UMAT to column 72.

Thank you very much in advance,
Nicolò

**Reply by jessecarterMOOSE** [reply]
I'd also like to know how to set compiler flags when compiling Fortran UMAT's.

For instance I'd like to compile a Fortran UMAT with `-O0`. If I compile my app with `METHOD=dbg make -n`, the C++ files get compiled with `-O0` but the Fortran still gets compiled with `-O2`. If I manually set `FFLAGS='-O0'` (as well as `FCFLAGS`, `F90FLAGS`, `F77FLAGS` because why not), I still get `-O2`.

@dschwen @GiudGiud

**Reply by GiudGiud** [reply]
See your own comment on this thread https://github.com/idaholab/moose/discussions/21476
about the PLUGIN_FLAGS. 
I'm documenting this to facilitate future searches

**Reply by jessecarterMOOSE** [reply]
Good call. Forgot that I answered my own question.

**Reply by grld** [reply]
Hello, thank you very much for your explanations. I followed the comments here and in this thread #21201. I am also trying to use a Fortran UMAT  in MOOSE. I struggled to compile but added three compiler flags [here](https://github.com/idaholab/moose/blob/next/framework/build.mk#L310). Now gfortran can compile the UMAT files. 
However, my UMAT consists of different .f files and .mods files. As @ngrilli has described in #21201, the files depend on each other and need to be compiled in a certain order. My problem is exactly there: The build.mk compiles the fortran files in an alphabetical order and not in the order I need. 
I tried to understand the Makefile @dschwen pointed out above. However, I don't understand to which Makefile I need to add the lines.

Sorry for my confusion and thank you very much in advance!
Jan

**Reply by YaqiWang** [reply]
Possibly we can manually add dependencies like
```
$(YOUR_FORTRAN_SRC_DIR)/A.f90: $(YOUR_FORTRAN_SRC_DIR)/B.$(obj-suffix)
```
In this way, `B.f90` will be compiled ahead of `A.f90`.

**Reply by grld** [reply]
Thank you @YaqiWang this would be a quick fix.
